### DISCUSSION TITLE
how to use SecFld.array() correctly

### DISCUSSION THREAD

**Discussion by devifaustine** [lschoe/mpyc#84] [Q&A] [answered]
I've been having trouble creating a secure field array these past few days. I want to implement operations done in bytes and the only way I see in the demos is to use an array in a secure field. But the program never terminates. Consider the following program, which receives two strings as inputs, converts them into an array in a secure field, and outputs the secure objects created. This program is stuck in the loop. I thought it had something to do with numpy instances used, and I tried using different numpy instances, but this still does not work. Am I missing something? Originally I wanted to concatenate these two bytes, but even outputting those bytes is impossible. I appreciate any help you can provide. 

```
import numpy as np
async def main():
    secfld = mpc.SecFld(2)
    await mpc.start()

    in_ = input("Give your input: ")
    # encode() converts string to bytes

    in_bit = ''.join(format(ord(i), '08b') for i in in_)
    print(in_bit)

    in_bit_list = [int(i) for i in in_bit]
    print(in_bit_list)

    inputs = mpc.input(secfld.array(np.array(in_bit_list)))

    print("input: ", await mpc.output(inputs))

    await mpc.shutdown()
```

**Comment by lschoe** [comment]
If I run the program with one party, there is no problem. So, I think you mean you experience this problem if you run the program with more than one party? For instance, with `-M 2`? 

In that case, you may add `--output-windows` on the command line, to see the other party. For that party also an input needs to be entered, otherwise there will be no progress.

**Reply by devifaustine** [reply]
Thanks for the fast reply. 
Yes, the problem occurs with -M2. I have already given the inputs in the command line for both parties using two different terminals. 
![image](https://github.com/lschoe/mpyc/assets/64829752/dff26039-3826-45cf-af30-73bc65eed25c)
And the program just stops there without terminating. 

When I tried to debug it, it outputs runtime error and behaves differently than when I'm just running it normally. 

```
Traceback (most recent call last):
  File "/usr/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
python-BaseException  File "/home/devifaustine/.local/lib/python3.10/site-packages/mpyc/asyncoro.py", line 431, in <lambda>

    task.add_done_callback(lambda t: _reconcile(decl, t))
  File "/home/devifaustine/.local/lib/python3.10/site-packages/mpyc/asyncoro.py", line 355, in _reconcile
    givn = task.result()
  File "/home/devifaustine/.local/lib/python3.10/site-packages/mpyc/asyncoro.py", line 283, in _wrap_in_coro
    return await awaitable
  File "/home/devifaustine/.local/lib/python3.10/site-packages/mpyc/asyncoro.py", line 271, in __await__
    val = self.coro.send(None)
  File "/home/devifaustine/.local/lib/python3.10/site-packages/mpyc/runtime.py", line 452, in _distribute
    y = [[field.array(unmarshal(r), check=False).reshape(shape) for r in shares]]
  File "/home/devifaustine/.local/lib/python3.10/site-packages/mpyc/runtime.py", line 452, in <listcomp>
    y = [[field.array(unmarshal(r), check=False).reshape(shape) for r in shares]]
  File "/home/devifaustine/.local/lib/python3.10/site-packages/mpyc/finfields.py", line 1289, in reshape
    return type(self)(self.value.reshape(*args, **kwargs), check=False)
ValueError: cannot reshape array of size 8 into shape (16,)
Traceback (enclosing MPyC coroutine call):
  File "/home/devifaustine/Documents/Post-Quantum-Blind-Signature-with-MPC/test_secobj.py", line 237, in main11
    print("input: ", await mpc.output(inputs))
  File "/home/devifaustine/.local/lib/python3.10/site-packages/mpyc/runtime.py", line 396, in input
    return y
Traceback (most recent call last):
  File "/snap/pycharm-community/364/plugins/python-ce/helpers/pydev/pydevd.py", line 1534, in _exec
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "/snap/pycharm-community/364/plugins/python-ce/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/home/devifaustine/Documents/Post-Quantum-Blind-Signature-with-MPC/test_secobj.py", line 463, in <module>
    mpc.run(main11())
  File "/home/devifaustine/.local/lib/python3.10/site-packages/mpyc/runtime.py", line 181, in run
    return self._loop.run_until_complete(f)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 647, in run_until_complete
    raise RuntimeError('Event loop stopped before Future completed.')
RuntimeError: Event loop stopped before Future completed.

Process finished with exit code 1
```

**Comment by lschoe** [comment]
Right, I see. Both parties input a string of possibly different lengths, but they don't know the input length used by the other party. To fix this you can set a maximum length for both parties, and let each party pad its input string `in_` to this maximum length by concatenating "` `"s for instance. Then all parties will end up using a secure array of the same length (hence of the same shape), which is a prerequisite for `mpc.input()` to work with arrays.

Another thing is that `mpc.output()` currently does not support *lists* of secure arrays. So, to see everything in `inputs`, you can put this `print("input: ", i, await mpc.output(inputs[i]))` in a `for i in range(len(inputs))` loop.

**Comment by devifaustine** [comment]
I tried padding it with blankspace and use a for loop for the mpc.output(), but for some reasons the input from the second party cannot be accessed and it just stops after printing out `input[0]`. 
![Screenshot from 2024-02-01 13-06-55](https://github.com/lschoe/mpyc/assets/64829752/b78f5425-d073-461c-8c8a-efeb2358172e)
Even if I change the for loop to just `print("input 1: ", await mpc.output(input[1]))` the program does not execute this. 
```
async def main():
    secfld = mpc.SecFld(2)
    await mpc.start()

    in_ = input("Give your input: ")
    
    # pad the input to max length 
    max = 10
    if len(in_) < max:
        in_ = pad(in_, max)

    in_bit = ''.join(format(ord(i), '08b') for i in in_)
    print(in_bit)

    in_bit_list = [int(i) for i in in_bit]
    print(in_bit_list)

    inputs = mpc.input(secfld.array(np.array(in_bit_list)))

    # mpc.output() does not support outputting list
    for i in range(len(inputs)):
        print("input ", i, ": ", await mpc.output(inputs[i]))

    await mpc.shutdown()

def pad(string, length):
    while len(string) < length:
        string += " "
    return string
```
Thanks again and sorry to bother you :)

**Comment by lschoe** [comment, marked as answer]
Your program works for me, but when I run it with MPyC v0.9 the same issue arises.

So, if you update to the latest version, e.g., using `pip install git+https://github.com/lschoe/mpyc` the problem will probably be gone.

**Comment by devifaustine** [comment]
Yeah, that seemed to be the problem :D, thank you so much for your help.
